EIPS/eip-2126.md

```diff
@@ -0,0 +1,131 @@
+| 0x01           | [Invalid](#invalid) |
+| 0x02           | [EIP712](#eip712)   |
+| 0x03           | [EthSign](#ethsign) |
+| 0x04           | [Wallet](#wallet)   |
+
+### Illegal
+
+The is the default value of the signature byte. A transaction that includes an Illegal signature will be reverted. Therefore, users must explicitly specify a valid signature type.
+
+### Invalid
+
+An `Invalid` signature always returns false. An invalid signature can always be recreated and is therefore offered explicitly. This signature type is largely used for testing purposes.
+
+### EIP712
+
+An [`EIP712`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md) signature is considered valid if the address recovered from calling ecrecover with the given hash and decoded `v`, `r`, `s` values is the same as the specified signer. In this case, the signature is encoded in the following way:
+
+| Offset | Length | Contents            |
+| ------ | ------ | ------------------- |
+| 0x00   | 1      | v (always 27 or 28) |
+| 0x01   | 32     | r                   |
+| 0x21   | 32     | s                   |
+
+### EthSign
+
+An `EthSign` signature is considered valid if the address recovered from calling ecrecover with the an EthSign-prefixed hash and decoded `v`, `r`, `s` values is the same as the specified signer.
+
+The prefixed `msgHash` is calculated with:
+
+```
+string constant ETH_PERSONAL_MESSAGE = "\x19Ethereum Signed Message:\n32";
+bytes32 msgHash = keccak256(abi.encodePacked(ETH_PERSONAL_MESSAGE, hash));
+```
+
+`v`, `r`, and `s` are encoded in the signature byte array using the same scheme as [EIP712 signatures](#EIP712).
+
+### Wallet
+
+The `Wallet` signature type allows an EOA account to sign messages on behalf of a smart contract account.
+
+As per [EIP-1271](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md), smart contract accounts can validate signatures using the public method `isValidSignature`. This method has the following interface:
+
+```typescript=
+// @dev Verifies that a signature is valid.
+// @param hash Message hash that is signed.
+// @param signerAddress Address of signer.
+// @param signature Proof of signing.
+// @return Validity of order signature.
+function isValidSignature(
+    bytes32 hash,
+    address signerAddress,
+    bytes memory signature
+)
+    public
+    view
+    returns (bool isValid);
+```
+
+## Web3 JSON API support
+
+We suggest appending an _optional_ parameter to existing Web3 signature functions. This parameter would contain key-value pairs. This is so that:
+
+- Web3 clients can choose when and where to include a signature type byte.
+- Existing code using these methods doesn't need to change.
+- Additional options can easily be added in future.
+
+Example using `web3.js`:
+
+```js
+web3.eth.personal.sign(
+    "...data...",
+    "0x...",
+    "password",
+    null, // callback
+    {
+        includeType: true, // informs signer to include signature type
+    }
+})
+```
+
```

Comment:
Missing the Rationale section.

Reply:
Would renaming the `Motivation` section to `Rationale` suffice?